# Data understanding

## 1. De dataset

In [1]:
import pandas as pd 
data = pd.read_csv("sap_storing_data_hu_project.csv") 

/Users/jet/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3146: DtypeWarning: Columns (0,11,19,21,24,25,27,32,33,36,37,38,39,40,46,47,48,49,51,53,55,56,60,73,75,78,80,82,83,86,87,91,95,99,102,103,104,107,108,109,110,111,112,113,116,118,119,120,121,122,123,124,129) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [2]:
data.shape

(898526, 140)

Wat meteen opvalt is dat er een groot aantal kolommen zijn: namelijk 140. Dit zijn dus 140 variabelen die we mogelijk mee moeten nemen verder in ons project. Bij deze stap in het project willen we gaan kijken welke variabelen kansrijk zijn om bijvoorbeeld als feature- of targetvariabele aangewezen kunnen worden. 

In eerste instantie maakt de dataset een redelijk rommelige indruk, dit komt waarschijnlijk door de lange en ietwat verwarrende namen van de kolommen en de verschillende datatypes waarin de waardes worden weergegeven. Het kan handig zijn om meer overzicht te creeëren binnen de dataset, door bijvoorbeeld kolommen een helderdere naam te geven. Gelukkig hebben we wel een legenda gekregen die ons uitlegt wat elke kolom in de dataset betekent. 


In [3]:
data.head()

Unnamed: 0  #stm_sap_meldnr  stm_mon_nr stm_vl_post     stm_sap_meld_ddt  \
0          0                0         NaN         NaN                  NaN   
1          1         50053211         0.0         NaN  02/01/2006 09:00:00   
2          2         50053213        48.0          GN  02/01/2006 12:35:00   
3          3         50053214        72.0          ZL  02/01/2006 16:40:00   
4          4         50053215        96.0          ZL  02/01/2006 22:30:00   

  stm_sap_mon_meld_ddt                             stm_sap_meldtekst_lang  \
0                  NaN                                                NaN   
1  02/01/2006 09:00:00   Logboeknr  Geeltje  :   49 Tijd: 0900   VL-Po...   
2  02/01/2006 12:35:00   Logboeknr  RBV  :   48 Tijd: 1235   VL-Post: ...   
3  02/01/2006 16:40:00   Logboeknr  RBV  :   72 Tijd: 1640   VL-Post: ...   
4  02/01/2006 22:30:00   Logboeknr  RBV  :   96 Tijd: 2230   VL-Post: ...   

     stm_mon_begin_ddt                         stm_sap_meldtekst  \
0                  NaN                                       NaN   
1                  NaN     Logboeknr  Geeltje  :   49 Tijd: 0900   
2  02/01/2006 12:35:00  Logboeknr  RBV  :   48 Tijd: 1235   VL-P   
3  02/01/2006 16:40:00  Logboeknr  RBV  :   72 Tijd: 1640   VL-P   
4  02/01/2006 22:30:00  Logboeknr  RBV  :   96 Tijd: 2230   VL-P   

  stm_mon_toelichting_trdl  ... stm_rapportage_maand stm_rapportage_jaar  \
0                      NaN  ...                    5                2014   
1                      NaN  ...                    1                2006   
2                      NaN  ...                    1                2006   
3                      NaN  ...                    1                2006   
4                      NaN  ...                    1                2006   

    stm_x_bron_publ_dt                            stm_x_bron_bestandsnaam  \
0  07/05/2014 05:30:12  \\PUHAPS0149\Informatica_Prod\Inbox\Informatic...   
1  02/01/2006 09:00:00  \\PUHAPS0149\Informatica_Prod\Inbox\Informatic...   
2  02/01/2006 12:35:00  \\PUHAPS0149\Informatica_Prod\Inbox\Informatic...   
3  02/01/2006 16:40:00  \\PUHAPS0149\Informatica_Prod\Inbox\Informatic...   
4  02/01/2006 22:30:00  \\PUHAPS0149\Informatica_Prod\Inbox\Informatic...   

    stm_x_bron_arch_dt stm_x_actueel_ind  stm_x_run_id  stm_x_bk  \
0  31/12/9999 00:00:00                 1         34415   3617016   
1  31/12/9999 00:00:00                 1          1518  12704590   
2  31/12/9999 00:00:00                 1          1518  12704591   
3  31/12/9999 00:00:00                 1          1518  12704592   
4  31/12/9999 00:00:00                 1          1518  12704593   

   stm_x_start_sessie_dt stm_x_vervallen_ind  
0    07/05/2014 05:44:24                   0  
1    13/06/2013 13:37:52                   0  
2    13/06/2013 13:37:52                   0  
3    13/06/2013 13:37:52                   0  
4    13/06/2013 13:37:52                   0  

[5 rows x 140 columns]

## 2. Missing values 

We willen onderzoeken hoeveel null waardes er zijn binnen de dataset. Op basis van het aantal null waardes kunnen we gaan beslissen welke kolommen te weinig waardes bevatten en dus niet bruikbaar zijn voor het project. 

In [4]:
notnull = data.notnull().sum().sum()
null = data.isnull().sum().sum()
sum = 87436170 + 38357470
nulldata = (null/sum)*100
nulldata 

30.4923762441408

Ongeveer 30% van de waardes in de dataset zijn leeg, als we uitzoeken welke kolommen dit vooral zijn kunnen we die kolommen droppen en zo het gehalte van lege waardes verminderen.

## 3. Correlatie

We willen de samenhang van de verschillende variabelen bekijken. Als variabelen sterk gecorreleert met elkaar zijn betekend dit dat er een verband te stellen is tussen die variabelen en ze dus mogelijk afhankelijk van elkaar fungeren. 

In [5]:
data.corr()

#stm_sap_meldnr  stm_mon_nr  \
#stm_sap_meldnr                   1.000000   -0.561259   
stm_mon_nr                       -0.561259    1.000000   
stm_geo_mld_uit_functiepl         0.224076   -0.207973   
stm_equipm_nr_mld                -0.203350    0.092692   
stm_km_van_mld                    0.007434   -0.009408   
stm_km_tot_mld                   -0.097676    0.114231   
stm_prioriteit                   -0.716977    0.348891   
stm_mon_nr__statuscode            0.154912   -0.365677   
stm_geo_gst_uit_functiepl         0.218851   -0.196783   
stm_equipm_nr_gst                -0.242909    0.140240   
stm_km_van_gst                    0.046235   -0.048957   
stm_km_tot_gst                   -0.010300    0.022103   
stm_oorz_code                     0.353319   -0.222087   
stm_fh_status                    -0.217210    0.015735   
stm_tao_telling_mutatie          -0.300278    0.328161   
stm_contractgeb_mld               0.068555   -0.084193   
stm_contractgeb_gst               0.084121   -0.078367   
stm_arbeid                        0.041034   -0.086499   
stm_progfh_gw_lwd_datum          -0.262559    0.169060   
stm_progfh_gw_teller              0.085909    0.018158   
stm_fh_duur                       0.065408   -0.059769   
stm_reactie_duur                  0.074173   -0.000121   
stm_status_sapnaarmon            -0.154282    0.420675   
stm_schadenr                      0.000158   -0.000888   
stm_rapportage_maand             -0.023604    0.011310   
stm_rapportage_jaar              -0.176788    0.402291   
stm_x_actueel_ind                 0.050593   -0.208674   
stm_x_run_id                     -0.141028    0.334218   
stm_x_bk                          0.157436   -0.344456   
stm_x_vervallen_ind              -0.254859    0.184754   

                           stm_geo_mld_uit_functiepl  stm_equipm_nr_mld  \
#stm_sap_meldnr                             0.224076          -0.203350   
stm_mon_nr                                 -0.207973           0.092692   
stm_geo_mld_uit_functiepl                   1.000000           0.048381   
stm_equipm_nr_mld                           0.048381           1.000000   
stm_km_van_mld                              0.044879           0.008139   
stm_km_tot_mld                              0.029848           0.115466   
stm_prioriteit                             -0.136484           0.228139   
stm_mon_nr__statuscode                      0.019642          -0.182107   
stm_geo_gst_uit_functiepl                   0.972265           0.023451   
stm_equipm_nr_gst                           0.009556           0.906940   
stm_km_van_gst                              0.086167          -0.011530   
stm_km_tot_gst                              0.050526           0.059147   
stm_oorz_code                               0.203407          -0.035554   
stm_fh_status                               0.005319          -0.012667   
stm_tao_telling_mutatie                    -0.135645          -0.036887   
stm_contractgeb_mld                         0.040836           0.166860   
stm_contractgeb_gst                         0.002255           0.167610   
stm_arbeid                                  0.035203          -0.037223   
stm_progfh_gw_lwd_datum                     0.001447           0.047060   
stm_progfh_gw_teller                       -0.003545           0.002489   
stm_fh_duur                                 0.023062           0.007824   
stm_reactie_duur                            0.030707           0.069651   
stm_status_sapnaarmon                      -0.040085           0.202643   
stm_schadenr                                0.000203          -0.011145   
stm_rapportage_maand                       -0.013678          -0.007791   
stm_rapportage_jaar                        -0.062358           0.353960   
stm_x_actueel_ind                           0.056368          -0.112766   
stm_x_run_id                               -0.056871           0.246999   
stm_x_bk                                    0.060315 

We tonen een klein overzicht van de variabelen die het sterkst gecorreleert zijn met elkaar. De variabelen die het sterkst met elkaar gecorreleert zijn, zoals bijvoorbeeld stm_progfh_gw_lwd_datum en stm_rapportage_jaar zeggen ons niet heel veel. Het is best logisch dat de prognose datum en het jaar waarin de rapportage gedaan is sterk met elkaar correleren. We zijn door deze correlaties dus niet perse veel wijzer geworden.

In [6]:
data.corr().unstack().sort_values().drop_duplicates()

stm_rapportage_jaar        stm_x_bk                    -0.820686
#stm_sap_meldnr            stm_prioriteit              -0.716977
stm_status_sapnaarmon      stm_mon_nr__statuscode      -0.716439
stm_x_bk                   stm_x_run_id                -0.596926
#stm_sap_meldnr            stm_mon_nr                  -0.561259
                                                          ...   
stm_progfh_gw_lwd_datum    stm_rapportage_jaar          0.912118
stm_contractgeb_mld        stm_contractgeb_gst          0.965901
stm_geo_mld_uit_functiepl  stm_geo_gst_uit_functiepl    0.972265
#stm_sap_meldnr            #stm_sap_meldnr              1.000000
stm_arbeid                 stm_progfh_gw_lwd_datum           NaN
Length: 435, dtype: float64

## 4. Dubbele data


Er zijn een aantal kolommen die identieke waardes hebben en dus redundant zijn. Zo zien we hieronder dat er twee overbodige kolommen zijn die samen dezelfde informatie weergeven als stm_sap_meld_ddt. Ook zien we dat stm_geo_gst en stm_geo_mld precies dezelfde informatie bevatten. Dit is allemaal onnodige dubbele informatie die we in de volgende stap data preparation willen gaan verwijderen. 

In [7]:
data[['stm_sap_meld_ddt', 'stm_sap_meldtijd', 'stm_sap_melddatum']]

stm_sap_meld_ddt stm_sap_meldtijd stm_sap_melddatum
0                       NaN              NaN               NaN
1       02/01/2006 09:00:00         09:00:00        02/01/2006
2       02/01/2006 12:35:00         12:35:00        02/01/2006
3       02/01/2006 16:40:00         16:40:00        02/01/2006
4       02/01/2006 22:30:00         22:30:00        02/01/2006
...                     ...              ...               ...
898521  11/05/2013 08:06:00         08:06:00        11/05/2013
898522  11/05/2013 09:21:00         09:21:00        11/05/2013
898523  11/05/2013 09:21:00         09:21:00        11/05/2013
898524  20/08/2016 14:15:17         14:15:17        20/08/2016
898525  20/08/2016 14:15:17         14:15:17        20/08/2016

[898526 rows x 3 columns]

In [8]:
data[['stm_geo_gst', 'stm_geo_mld']]

stm_geo_gst stm_geo_mld
0              NaN         NaN
1              624         624
2              201         201
3               25          25
4               12          12
...            ...         ...
898521         NaN         560
898522         468         468
898523         468         468
898524         102         102
898525         102         102

[898526 rows x 2 columns]

In de volgende stap data preparation gaan we de ontdekkingen verwerken en zo zorgen dat we de juiste data hebben om in ons model te gaan gebruiken. 